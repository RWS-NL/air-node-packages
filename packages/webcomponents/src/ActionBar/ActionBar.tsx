import Grid from '@material-ui/core/Grid/Grid';
import css from './ActionBar.scss';
import Button from '../Button/Button';
import classnames from 'classnames';
import React, { FC, Fragment, ReactNode, useMemo, memo } from 'react';

export interface ActionBarProps {
  /** The title to show in the ActionBar */
  title: string | ReactNode;
  /** Whether an action button should be displayed in the ActionBar */
  shouldHaveButton?: boolean;
  /** Whether the button should be disabled, only relevant when `shouldHaveButton` is set to true */
  shouldDisableButton?: boolean;
  /** The text that should be shown in the button */
  buttonLabel?: string;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: string;
  /** Custom CSS classes to pass to the button */
  customclasses?: string | string[];
  /** The action that should be invoked when the button is clicked */
  buttonAction?: (...args: any[]) => any;
}

/**
 * Creates an action bar using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the actionbar
 * @example
 * <ActionBar title='Cool title' />
 */
export const ActionBar: FC<ActionBarProps> = props => {
  const getTitle = useMemo<ActionBarProps['title']>(() => {
    if (typeof props.title === 'string') {
      return (
        <h1 data-qa='action-bar-title' className={classnames(css.actionBarHeader)}>
          {props.title}
        </h1>
      );
    }

    if (typeof props.title === 'function') return props.title();

    return props.title;
  }, [props]);

  return (
    <div className={classnames('navigation-bar', css.actionBar)} data-qa={props['data-qa']}>
      <Grid container direction='row' justify='space-between' alignItems='center' style={{ height: '100%' }}>
        <Grid item key={1} xs={6}>
          {getTitle}
        </Grid>
        <Grid item key={2} xs={6}>
          {props.shouldHaveButton ?
            <Button
              data-qa='action-bar-button'
              variant='contained'
              color='primary'
              label={props.buttonLabel}
              onClick={() => props.buttonAction ? props.buttonAction() : null}
              disabled={props.shouldDisableButton}
            />
            : <Fragment />
          }
        </Grid>
      </Grid>
    </div>
  );
};

export default memo(ActionBar);