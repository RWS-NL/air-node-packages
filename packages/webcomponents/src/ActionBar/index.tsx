import { Box, BoxProps, Typography, TypographyProps } from '@material-ui/core';
import Grid from '@material-ui/core/Grid/Grid';
import clsx from 'clsx';
import React, { memo, ReactNode, useMemo } from 'react';
import { When } from 'react-if';
import { Button } from '../Button';
import css from './ActionBar.scss';

export interface ActionBarProps {
  /** The title to show in the ActionBar */
  title: ReactNode;
  /**
   * Whether an action button should be displayed in the ActionBar
   * @default false
   */
  shouldHaveButton?: boolean;
  /**
   * Whether the button should be disabled
   * Only relevant when `shouldHaveButton` is set to true
   * @default false
   */
  shouldDisableButton?: boolean;
  /** The text that should be shown in the button */
  buttonLabel?: string;

  /** Additional props to apply to the Typography component */
  TypographyProps?: TypographyProps;
  /** Additional props to pass to the encompasing Box component */
  BoxProps?: BoxProps;

  /** Data-qa tag to apply action bar bounding box */
  'data-qa'?: string;
  /** The action that should be invoked when the button is clicked */
  buttonAction?: (...args: any[]) => any;
}

/**
 * Creates an action bar using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the actionbar
 * @example
 * ```jsx
 * <ActionBar title='Cool title' />
 * ```
 */
export const ActionBar = memo(
  ({
    title,
    'data-qa': dataQa,
    shouldDisableButton = false,
    shouldHaveButton = false,
    buttonAction,
    buttonLabel,
    BoxProps,
    TypographyProps
  }: ActionBarProps) => {
    const getTitle = useMemo<ActionBarProps['title']>(() => {
      if (typeof title === 'string') {
        return (
          <Typography
            variant='h1'
            data-qa='action-bar-title'
            color='textPrimary'
            {...TypographyProps}
            className={clsx(css.actionHeader, TypographyProps?.className, TypographyProps?.classes)}
          >
            {title}
          </Typography>
        );
      }

      if (typeof title === 'function') return title();

      return title;
    }, [TypographyProps, title]);

    return (
      <Box component='div' data-qa={dataQa} {...BoxProps} className={clsx(css.actionBar, BoxProps?.className)}>
        <Grid container direction='row' justify='space-between' alignItems='center' className={css.actionGridLeft}>
          <Grid item key={1} xs={6}>
            {getTitle}
          </Grid>
          <When condition={Boolean(shouldHaveButton)}>
            <Grid item key={2} xs={5} sm={6} className={css.actionGridRight}>
              <Button
                data-qa='action-bar-button'
                variant='contained'
                color='primary'
                label={buttonLabel}
                onClick={() => (buttonAction ? buttonAction() : null)}
                disabled={Boolean(shouldDisableButton)}
              />
            </Grid>
          </When>
        </Grid>
      </Box>
    );
  }
);
