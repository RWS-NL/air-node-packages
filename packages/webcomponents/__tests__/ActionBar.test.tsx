import ActionBar from '../src/ActionBar/ActionBar';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

const mockCallback = jest.fn();

describe('Component Testing', () => {
  let actionBar: ShallowWrapper;

  describe('ActionBar with button', () => {
    test('confirm the button click is registered', () => {
      actionBar = shallow(<ActionBar title='42' buttonLabel='label' buttonAction={mockCallback} shouldHaveButton />);
      const actionButton = actionBar.find('[data-qa="action-bar-button"]');
      actionButton.simulate('click');
      expect(mockCallback).toHaveBeenCalledWith();
      expect(actionButton.dive().text()).toBe('label');
    });

    test('confirm the button click passes arguments', () => {
      const param = 42;

      actionBar = shallow(
        <ActionBar title='42' buttonLabel='label'
          buttonAction={() => mockCallback(param)}
          shouldHaveButton
        />
      );

      const actionButton = actionBar.find('[data-qa="action-bar-button"]');
      actionButton.simulate('click');
      expect(mockCallback).toHaveBeenCalledWith();
      expect(mockCallback).toHaveBeenCalledWith(param);
      expect(actionButton.dive().text()).toBe('label');
    });
  });

  describe('ActionBar without button', () => {
    beforeEach(() => {
      actionBar = shallow(<ActionBar title='42' />);
    });

    test('should show title', () => {
      const title = actionBar.find('[data-qa="action-bar-title"]');
      const actionButton = actionBar.find('[data-qa="action-bar-button"]');
      expect(title.text()).toBe('42');
      expect(actionButton).toHaveLength(0);
    });
  });

  describe('ActionBar with ReactNode function based title', () => {
    beforeEach(() => {
      actionBar = shallow(<ActionBar title={() => <div data-qa='action-bar-custom-title'>text</div>} />);
    });

    test('should show title node', () => {
      const title = actionBar.find('[data-qa="action-bar-custom-title"]');

      expect(title.text()).toBe('text');
    });
  });

  describe('ActionBar with ReactNode element based title', () => {
    beforeEach(() => {
      actionBar = shallow(<ActionBar title={<div data-qa='action-bar-custom-title'>text</div>} />);
    });

    test('should show title node', () => {
      const title = actionBar.find('[data-qa="action-bar-custom-title"]');

      expect(title.text()).toBe('text');
    });
  });
});


describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const actionBar = shallow(<ActionBar title='42' />);
    expect(actionBar).toMatchSnapshot();
  });

  test('With Button', () => {
    const actionBar = shallow(
      <ActionBar
        title={() => <div data-qa='action-bar-custom-title'>text</div>} buttonLabel='label'
        buttonAction={mockCallback} shouldHaveButton
      />
    );
    expect(actionBar).toMatchSnapshot();
  });

  test('AdditionalTest', () => {
    const actionBar = shallow(
      <ActionBar
        title={<div data-qa='action-bar-custom-title'>text</div>} buttonLabel='label'
        buttonAction={mockCallback} shouldHaveButton
        customclasses={'snapshot-class'} data-qa='snapshot-qa'
      />
    );
    expect(actionBar).toMatchSnapshot();
  });
});