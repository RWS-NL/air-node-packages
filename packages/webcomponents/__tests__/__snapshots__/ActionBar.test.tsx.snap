// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing AdditionalTest 1`] = `
<div
  className="navigation-bar actionBar"
  data-qa="snapshot-qa"
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    direction="row"
    justify="space-between"
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="1"
      xs={6}
    >
      <div
        data-qa="action-bar-custom-title"
      >
        text
      </div>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="2"
      xs={6}
    >
      <Memo(Button)
        color="primary"
        data-qa="action-bar-button"
        label="label"
        onClick={[Function]}
        variant="contained"
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</div>
`;

exports[`Snapshot Testing Required Props 1`] = `
<div
  className="navigation-bar actionBar"
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    direction="row"
    justify="space-between"
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="1"
      xs={6}
    >
      <h1
        className="actionBarHeader"
        data-qa="action-bar-title"
      >
        42
      </h1>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="2"
      xs={6}
    />
  </WithStyles(ForwardRef(Grid))>
</div>
`;

exports[`Snapshot Testing With Button 1`] = `
<div
  className="navigation-bar actionBar"
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    direction="row"
    justify="space-between"
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="1"
      xs={6}
    >
      <div
        data-qa="action-bar-custom-title"
      >
        text
      </div>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="2"
      xs={6}
    >
      <Memo(Button)
        color="primary"
        data-qa="action-bar-button"
        label="label"
        onClick={[Function]}
        variant="contained"
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</div>
`;
